src/Bundle/DependencyInjection/SyliusMailerExtension.php

```diff
@@ -66,37 +69,60 @@ public function getConfiguration(array $config, ContainerBuilder $container): Co
 
     private function configureSenderAdapter(ContainerBuilder $container, array $config): void
     {
-        $bundles = (array) $container->getParameter('kernel.bundles');
+        if (isset($config['sender_adapter'])) {
+            $container->setAlias('sylius.email_sender.adapter', $config['sender_adapter']);
+        }
 
-        $defaultSenderAdapter = 'sylius.email_sender.adapter.default';
-        if (array_key_exists('SwiftmailerBundle', $bundles)) {
-            $swiftmailerAdapter = new ChildDefinition('sylius.email_sender.adapter.abstract');
-            $swiftmailerAdapter->setClass(SwiftMailerAdapter::class);
-            $swiftmailerAdapter->setArguments([new Reference('mailer'), new Reference('event_dispatcher')]);
-            $swiftmailerAdapter->setPrivate(false);
+        if (!$container->hasAlias('sylius.email_sender.adapter')) {
+            $container->setAlias('sylius.email_sender.adapter', 'sylius.email_sender.adapter.default');
+        }
+    }
 
-            $container->setDefinition('sylius.email_sender.adapter.swiftmailer', $swiftmailerAdapter);
-            $defaultSenderAdapter = 'sylius.email_sender.adapter.swiftmailer';
+    private function configureRendererAdapter(ContainerBuilder $container, array $config): void
+    {
+        if (isset($config['renderer_adapter'])) {
+            $container->setAlias('sylius.email_renderer.adapter', $config['renderer_adapter']);
         }
 
-        $container->setAlias('sylius.email_sender.adapter', $config['sender_adapter'] ?? $defaultSenderAdapter);
+        if (!$container->hasAlias('sylius.email_renderer.adapter')) {
+            $container->setAlias('sylius.email_renderer.adapter', 'sylius.email_renderer.adapter.default');
+        }
     }
 
-    private function configureRendererAdapter(ContainerBuilder $container, array $config): void
+    private function preConfigure(ContainerBuilder $container): void
     {
-        $bundles = (array) $container->getParameter('kernel.bundles');
+        /** @var array<string, class-string<BundleInterface>> $bundles */
+        $bundles = $container->getParameter('kernel.bundles');
+
+        if ($container->has('mailer.mailer')) {
+            $symfonyMailerAdapter = new ChildDefinition('sylius.email_sender.adapter.abstract');
+            $symfonyMailerAdapter->setClass(SymfonyMailerAdapter::class);
+            $symfonyMailerAdapter->setArguments([new Reference('mailer.mailer'), new Reference('event_dispatcher')]);
+            $symfonyMailerAdapter->setPublic(true);
+
+            $container->setDefinition('sylius.email_sender.adapter.symfony_mailer', $symfonyMailerAdapter);
+            $container->setAlias('sylius.email_sender.adapter', 'sylius.email_sender.adapter.symfony_mailer');
+        }
```

Comment:
The problem with this block is it does not work :dancer: 😄 I've just checked it on Sylius and in [this line](https://github.com/Sylius/SyliusMailerBundle/pull/102/files#diff-a57a7416c3f7616a0cd74e1053206497173323e871ac5b569184b8fdc7a8dde9R97) we still do not have the full services list available (only those from `services.xml` are loaded, [line 38](https://github.com/Sylius/SyliusMailerBundle/pull/102/files#diff-a57a7416c3f7616a0cd74e1053206497173323e871ac5b569184b8fdc7a8dde9R38)). So it never registers `sylius.email_sender.adapter.symfony_mailer` and when it's used in the configuration we got an error 💥 It's not a problem with swiftmailer, as to register it we check is the bundle registered, which is always a static list of FQCNs.

My first idea to fix this problem is to move these adapters' registration into the compiler pass. There we will have all the services registered properly and can base on them to register (or not) the adapters. I've already checked that :D and it seems to work. We can even unify the implementation and do not check for the existence of `SwiftmailerBundle` to register a swiftmailer adapter, but to check if `swiftmailer.mailer.default` is available.

It works in tests, as we mock the service existence there 🐎 It probably indicated we need to think about some more reliable tests for this bundle rather than only unit ones 🖖

Reply:
The other thing is, that I would keep `SwiftMailerAdapter` as the default one to keep the backward compatibility. We can still deprecate it and inform the developers to move to SymfonyMailer 🔥